src/lib.rs

```diff
@@ -109,60 +109,21 @@ where
 /// ```
 pub fn decode_polyline(polyline: &str, precision: u32) -> Result<LineString<f64>, String> {
     let mut index = 0;
-    let mut at_index;
     let mut lat: i64 = 0;
     let mut lng: i64 = 0;
     let mut coordinates = vec![];
     let base: i32 = 10;
     let factor = i64::from(base.pow(precision));
 
-    while index < polyline.len() {
-        let mut shift = 0;
-        let mut result = 0;
-        let mut byte;
-
-        loop {
-            at_index = polyline
-                .chars()
-                .nth(index)
-                .ok_or("Couldn't decode Polyline")?;
-            byte = at_index as u64 - 63;
-            index += 1;
-            result |= (byte & 0x1f) << shift;
-            shift += 5;
-            if byte < 0x20 {
-                break;
-            }
-        }
+    let chars = polyline.as_bytes();
 
-        let latitude_change: i64 = if (result & 1) > 0 {
-            !(result >> 1)
-        } else {
-            result >> 1
-        } as i64;
-
-        shift = 0;
-        result = 0;
-
-        loop {
-            at_index = polyline
-                .chars()
-                .nth(index)
-                .ok_or("Couldn't decode Polyline")?;
-            byte = at_index as u64 - 63;
-            index += 1;
-            result |= (byte & 0x1f) << shift;
-            shift += 5;
-            if byte < 0x20 {
-                break;
-            }
+    while index < chars.len() {
+        let (latitude_change, new_index) = trans(&chars, index);
+        if index >= chars.len() {
```

Comment:
```suggestion
        if new_index >= chars.len() {
```
I see no point to check `index` in this line, maybe `new_index` should be checked instead?